### ISSUE TITLE
Configurable request timeout

### ISSUE THREAD

**Issue by nnevala** [meteor/meteor#3826]
This issue is two-fold:

**Firstly**, it would be great if the server side connection and response timeout values (set by the WebApp) were configurable somehow (maybe read them from an environment variable, or a settings file?). Requests to server-side routes now time out after two minutes, and this is posing an issue for me now.

I get that for typical Meteor applications, running under typical conditions two minutes is plenty, but in some circumstances this isn't enough.

**Secondly**, but very much related to this issue, I've been poking around in the `webapp` package trying to figure out a way to simply hack around this issue. What turned out is that the [`LONG_SOCKET_TIMEOUT` value defined in `packages/webapp/webapp_server.js`](https://github.com/meteor/meteor/blob/d4d349ca96b57f4cbf36d84b41bcb3ca5bd70850/packages/webapp/webapp_server.js#L18) doesn't actually work if you set a timeout value greater than 120 seconds.

In other words, at the two minute mark something else seems to forcefully close the socket, ignoring the `LONG_SOCKET_TIMEOUT`. Setting a timeout value less than 120 seconds seems to work correctly, however.

Any ideas what might be closing the connection? I might at the very least be able to hack my way from there to get to a temporary solution for this issue. Furthermore,  I would like to provide a repository demonstrating these issues, but I couldn't figure out a way to run meteor with a custom WebApp package.

**Comment by glasser** [comment]
If you want to override the webapp package, just copy it into the `packages` directory of your app.

**Comment by nnevala** [comment]
Hi @glasser, and thanks for the tip! Easy as pie.

I set up a repository demonstrating the issue with the `LONG_SOCKET_TIMEOUT` value handling here: https://github.com/nnevala/meteor-timeout-bug

**Comment by glasser** [comment]
When you run `meteor` in development mode, it runs an extra HTTP proxy in addition to the actual app.  This lets us do things like display build errors in the screen, and lets you go to the app port in your browser before the app is fully built.  If you run meteor with `meteor --app-port 3040` you can specify what port should be used for the actual app process (the normal `--port` specifies the proxy port).

I tested your reproduction (well done, thanks!) and saw that when I curl'd against the proxy port, the behavior you saw happened; but when I curl'd against the app port, the extra 120 second timeout didn't happen.  I think this means that the 120 second timeout comes from the proxy.

And in fact, I think it's the default Node 0.10 http server timeout: https://github.com/joyent/node/blob/7edfd5f6d11f7a219408bc1edf265dbfe2b90900/lib/http.js#L1884

So I think the timeouts specified in webapp work, but if you are running other proxies in front of your meteor server it may affect observed behavior!

**Comment by terraswat** [comment]
So how does one change the timeout period for the development proxy?

**Comment by lobbin** [comment]
So, it this solution this usable on the latest Meteor, or have there been any improvements?

**Comment by leomelin** [comment]
This issue was still bugging me. I couldn't find any other solution to fix this than to search and replace SHORT_SOCKET_TIMEOUT and LONG_SOCKET_TIMEOUT in production build webapp.js to have longer timeout value. I have no idea if newer meteor version has fixed this already? I am using meteor 1.3.14 in production.

**Comment by involve-it** [comment]
To fix Webapp's LONG_SOCKET_TIMEOUT issue without cloning locally package, add 1 line: 
_WebApp.connectHandlers.use('/hello', (req, res, next) => {
**req.setTimeout(600000);**
..._ in your webapp's '_use_' callback method.
And yes, this SHOULD be configurable. We are not hackers, we a developers, consuming this library.

**Comment by wlodi83** [comment]
@involve-it,

it does not work either. I've tested it using following code example:

```
     app.get('/test', async function (req, res) {
        req.setTimeout(150 * 1000);
        console.log('before', new Date().toString());
        let p = new Promise(resolve =>
            setTimeout(() => {
                console.log('finished', new Date().toString());
                resolve(true);
            }, 150 * 1000)
        );
        await p;
        console.log('after', new Date().toString());

        return res.send(`all good ${new Date().toString()}`);
    });
```

Meteor server drops connection after 2 minutes.

**Comment by chfritz** [comment]
Surprisingly, what worked for me was to add a handler for the `timeout` event on the response object.
```
app.get('/debug/timeout', (req, res) => {
  console.log('let\'s wait');
  res.on('timeout', () => console.log('res timeout: ignore'));  // without this it times out after 2 minutes, with it, it doesn't
  setTimeout(() => res.send('done'), 140 * 1000);
});
```
However, this only works when connecting to the app-port directly, as suggested by @glasser . And yes, it was necessary, at least in my case, when connecting directly to that port (bypassing the proxy).

**Comment by davidworkman9** [comment]
Leaving this here for anyone who stumbles across it, I tried a few of the examples above and it was only this:
```res.on('timeout', () => console.log('res timeout: ignore'));```

That seemed to work. Reading closer I realized `req.setTimeout` doesn't work because it's actually the response that I needed to increase the timeout on, so for me the fix was something like:
**res**.setTimeout(600000);
